### PR TITLE
node: topologymgr: Graduate Kubelet Topology Manager to GA 

### PR DESCRIPTION
**Pull request by swatisehgal**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Captures changes to graduate Kubelet Topology Manager to GA.

#### Which issue(s) this PR fixes:

Issue: [#693](https://github.com/kubernetes/enhancements/issues/693)

#### Does this PR introduce a user-facing change?

```release-note
Graduate Kubelet Topology Manager to GA.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
KEP: https://github.com/kubernetes/enhancements/pull/3745
```

### CODE REVIEW

**Review by ffromani — COMMENTED** [review]
minor comments inside.
LGTM once the dependencies are merged.

**Comment by ffromani** [comment]
ok, according to https://github.com/kubernetes/enhancements/issues/693 this seems the last PR needed (and this seems correct to me). So let's discuss the inline comments and should be fine.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (swatisehgal) [reply pkg/kubelet/cm/container_manager_linux.go:300]: Yes, it is.

File: pkg/kubelet/cm/internal_container_lifecycle.go
Comment (swatisehgal) [reply pkg/kubelet/cm/internal_container_lifecycle.go:60]: Good point, will return the error directly.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (ffromani) [reply pkg/kubelet/cm/container_manager_linux.go:300]: great, then no further action is needed on this front.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (ffromani) [inline pkg/kubelet/cm/container_manager_linux.go:296]: sorry for late notice: should we rename these fields dropping the `Experimental` prefix? I think we should

/cc @SergeyKanzhelev

**Review by ffromani — COMMENTED** [review]
I apologize for the late notice, but it seems we missed a trivial rename. Besides that, everything else LGTM

If you agree with the rename, you can either push a separate commit with the rename or amend the existing last one

File: pkg/kubelet/cm/container_manager_linux.go
Comment (swatisehgal) [reply pkg/kubelet/cm/container_manager_linux.go:296]: No worries, I can add a commit for this.

**Comment by ffromani** [comment]
/lgtm

thanks!

**Comment by ffromani** [comment]
/retest

 pull-kubernetes-e2e-kind-ipv6 failures can only be flakes considering the content of this PR

**Comment by ffromani** [comment]
/retest

```
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration: TestChangeCRD expand_more
```
this looks like a flake

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Comment by swatisehgal** [comment]
/assign @klueska @mrunalp 
for approval on kubelet changes

/assign @dims @derekwaynecarr
for features approval

**Comment by swatisehgal** [comment]
/cc @pohly @bart0sh 
for input on changes related to removal of `resourceAllocator` as `cm.draManager` was passed into it.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (klueska) [inline pkg/kubelet/cm/container_manager_linux.go:702]: It was an oversight not to remove this from here in the first place because allocation was moved out of the pod admission loop and into container start before the DRA PR ever merged. This is leftover from intermediate changes that never merged.

**Review by klueska — COMMENTED** [review]
Did a quick pass, including addressing your comment about the draManager in the (now removed) resourceAllocator.

/lgtm
/approve

File: pkg/kubelet/cm/container_manager_linux.go
Comment (swatisehgal) [reply pkg/kubelet/cm/container_manager_linux.go:702]: Thanks @klueska for clarifying this!

**Comment by swatisehgal** [comment]
/label api-review
due to changes made in `apis/config/validation` .

File: cmd/kubelet/app/server.go
Comment (SergeyKanzhelev) [inline cmd/kubelet/app/server.go:718]: why it still refers to `features.TopologyManager`?

```suggestion
			return fmt.Errorf("topology manager policy options %v require feature gate  %q enabled",
				s.TopologyManagerPolicyOptions, features.TopologyManagerPolicyOptions)
```

File: pkg/kubelet/cm/container_manager.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/cm/container_manager.go:159]: this still requires `TopologyManagerPolicyOptions` gate? Should it stay Experimental?

File: cmd/kubelet/app/server.go
Comment (swatisehgal) [reply cmd/kubelet/app/server.go:718]: Good point, will remove.

File: pkg/kubelet/cm/container_manager.go
Comment (swatisehgal) [reply pkg/kubelet/cm/container_manager.go:159]: Yes it does. Rethinking this, I believe it makes sense to keep the experimental prefix for this as it is still an alpha feature thereby requiring a feature gate to enable it. Once the feature graduates to Beta, we can remove the prefix.

File: cmd/kubelet/app/server.go
Comment (ffromani) [inline cmd/kubelet/app/server.go:754]: yes, this seems the tight change at this point in time. ExperimentalTopologyManagerPolicyOptions is governed by a different FG.

**Review by ffromani — COMMENTED** [review]
/lgtm

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [inline pkg/features/kube_features.go:760]: Also need to update types.go - remove wording `requires the TopologyManager feature gate to be enabled`

File: test/e2e_node/topology_manager_test.go
Comment (SergeyKanzhelev) [inline test/e2e_node/topology_manager_test.go:202]: I think tests do not need the tags `[Feature:TopologyManager][NodeFeature:TopologyManager]` any longer

Comment (swatisehgal) [reply test/e2e_node/topology_manager_test.go:202]: A tag such as this can be useful if someone want to run just topology manager node e2e tests e.g. for debugging purposes. I will remove `[Feature:TopologyManager]` and we can keep just `[NodeFeature:TopologyManager]`.

File: pkg/features/kube_features.go
Comment (swatisehgal) [reply pkg/features/kube_features.go:760]: Sure, will do now. Thanks!

File: pkg/kubelet/cm/container_manager_linux.go
Comment (liggitt) [inline pkg/kubelet/cm/container_manager_linux.go:291]: looking inside the NewManager() impl, should the calls to NewPolicyOptions and NewNUMAInfo calls be skipped if the policy is PolicyNone? Currently, it looks like we can fail to start kubelets if we can't discover NUMA topology, even though that won't be used at all if the policy is PolicyNone

Comment (ffromani) [reply pkg/kubelet/cm/container_manager_linux.go:291]: This is a valid point. Previously users had the chance to exclude this flow with the FG, this option is now being removed. My 2c: this predates the GA graduation and could require some code massaging (e.g. not a simple if) so IMO is best addressed with a followup PR, which should be tracked (but probably not blocking) in the GA-graduation issue.  I'll let @klueska (cc @SergeyKanzhelev ) make the final call.

Comment (SergeyKanzhelev) [reply pkg/kubelet/cm/container_manager_linux.go:291]: https://github.com/kubernetes/kubernetes/issues/116391

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

Created an issue to follow up and minimize the number of calls needed on initialization when policy is none. Should not block GA

File: pkg/kubelet/cm/container_manager_linux.go
Comment (ffromani) [reply pkg/kubelet/cm/container_manager_linux.go:291]: thanks @SergeyKanzhelev !

**Pull request merged**
